uber/models.py

```diff
@@ -494,93 +494,138 @@ def lookup_attendee(self, full_name, email, zip_code):
-                    self.shift_badges(badge_type, badge_num + offset, up=True)
+                if attendee.badge_num:
+                    offset = 1 if old_badge_type == attendee.badge_type and attendee.badge_num > (old_badge_num or 0) else 0
+                    self.shift_badges(attendee.badge_type, attendee.badge_num + offset, up=True)
+                attendee.badge_num = badge_num_keep
+
+            if not attendee.badge_num and self.needs_badge_num(attendee):
+                attendee.badge_num = self.get_next_badge_num(attendee.badge_type)
 
-            attendee.badge_num = badge_num
-            attendee.badge_type = badge_type
             return 'Badge updated'
 
+        def auto_badge_num(self, badge_type):
+            """
+            Gets the next available badge number for a badge type's range.
+
+            This ignores the session so that plugins can override the logic here if need be without worrying about
+            handling dirty sessions.
+
+            :param badge_type: Used as a starting point if no badges of the same type exist, and to select badges within
+            a specific range.
+            :return:
+            """
+            sametype = self.query(Attendee.badge_num).filter(Attendee.badge_type == badge_type,
+                                                   Attendee.badge_num >= c.BADGE_RANGES[badge_type][0],
+                                                   Attendee.badge_num <= c.BADGE_RANGES[badge_type][1])
+            if sametype.count():
+                sametype_list = [int(row[0]) for row in sametype.order_by(Attendee.badge_num).all()]
+
+                # Searches badge range for a gap in badge numbers; if none found, returns the latest badge number + 1
+                first = 0
+                last = sametype.count() - 1
+                if sametype_list[last] - sametype_list[first] == last:
+                    return sametype.order_by(Attendee.badge_num.desc()).first().badge_num + 1
+                else:
+                    middle = int((first+last)/2)
+
+                    while first < last:
+                        if (sametype_list[middle]-sametype_list[first]) != (middle - first):
+                            if (middle-first) == 1 and (sametype_list[middle]-sametype_list[first] > 1):
+                                return sametype_list[middle] - 1
+                            last = middle
+                        elif (sametype_list[last]-sametype_list[middle]) != (last-middle):
+                            if (last-middle) == 1 and (sametype_list[last]-sametype_list[middle] > 1):
+                                return sametype_list[middle] + 1
+                            first = middle
+                        else:
+                            return sametype.order_by(Attendee.badge_num.desc()).first().badge_num + 1
+
+                        middle = int((first+last)/2)
+            else:
+                return c.BADGE_RANGES[badge_type][0]
```

Comment:
In the long run, if we do end up keeping this kind of "guard against the degenerate case where everything got messed up" then an alternate approach could be to maintain a lock-guarded "next badge num" for each type in memory, and have a background task update it.  That would allow us to do this kind of check in a background task instead of re-performing it every time we need to get the next badge number.

With that being said, this approach is simpler and probably the better way to go for now.